manifests/repos.pp

```diff
@@ -22,19 +37,62 @@
       $package_key   = $docker::package_key
       $package_repos = $docker::package_repos
 
+      if ( $facts['os']['distro']['id'] == 'Debian' and versioncmp($facts['os']['distro']['release']['major'],'10') >= 0 ) or ( $facts['os']['distro']['id'] == 'Ubuntu' and versioncmp($facts['os']['distro']['release']['major'],'22') >= 0 ) { # lint:ignore:140chars
+        # fix deprecated apt-key warnings
+        if $curl_ensure {
+          ensure_packages(['curl'])
+        }
+        if $gpg_ensure {
+          ensure_packages(['gpg'])
+        }
+        if $keyring_force_update {
+          exec { 'Remove Docker-GPG-Key':
+            path    => '/bin/',
+            cwd     => '/tmp',
+            command => "rm ${keyring}",
+          }
+          Exec['Remove Docker-GPG-Key'] -> Exec['Install Docker-GPG-Key']
+        }
+        exec { 'Install Docker-GPG-Key':
```

Comment:
Don't use exec to download something. Better use archive module and archive type.

Reply:
Generally I agree with you.  But for this part I copied manifests/compose.pp line 92-99 ( download of docker compose ) to keep in style with the rest of the module.  
I really don't know if I prefer to handle the same scenario in different ways.

Reply:
In addition, the downloaded file needs to be 'dearmored' by gpg, via exec I could simply pipe the download output to stdin of gpg.  
Do you have a code example how to achieve this with puppet/archive @tuxmea ?

Reply:
I think with puppet-archive, you could use `extract_command` to do the dearmoring.

Reply:
Untested example:
```puppet
archive { $keyring:
  ensure          => present,
  source          => "https://download.docker.com/linux/${docker::os_lc}/gpg"
  extract         => true,
  extract_command => "gpg --dearmor -o ${keyring}"
}
```

Reply:
The requested change was added via 177dfe6158f8bc2dbf4d623cf964d67463f64ea3.